### PR TITLE
Convert core-data selectors from jsDoc annotations into TypeScript type signatures

### PR DESCRIPTION
**Pull request by adamziel**
## What?

This is a subset of https://github.com/WordPress/gutenberg/pull/39025 – a mega branch that proposes TypeScript signatures for all `@wordpress/core-data` selectors.

This PR removes the jsDoc annotations like `@param {Object}` and `@return {string}` in favor of their TypeScript counterparts. This is the first step towards introducing exhaustive type signatures after https://github.com/WordPress/gutenberg/pull/40024 is merged.

There are two shortcomings for now:

1. ~The doc tool that handles `START TOKEN(Autogenerated selectors` fails. Apparently this is the first instance of mixing it with TypeScript files in the repo.~ (edit: should be fixed by #40236)
2. `createRegistrySelector` isn't typed at the moment, which means that calling it like `const mySelector = createRegistrySelector( select => ( state: State ) => null )` produces a `mySelector` variable of type `any`. We could solve it here by adding a `.d.ts` file to provide the missing signature, or we could solve this problem in `@wordpress/data` instead.

## Testing Instructions

Confirm the checks are green and that no entity configuration got changed when I split the large array into atomic declarations. The changes here should only affect the type system so there is nothing to test in the browser.

cc @dmsnell @jsnajdr @youknowriad @sarayourfriend @getdave @draganescu @scruffian

### CODE REVIEW

**Comment by dmsnell** [comment]
I'm confused by this: why create new types out of band instead of adding type annotation to the actual selectors file?

**Comment by adamziel** [comment]
@dmsnell I just started this one without realizing selectors.js can now be easily renamed to selectors.ts. I had a much harder time doing that last month :-) Let's rebase and apply these changes directly.

**Comment by adamziel** [comment]
Ah yes, our doc building tool throw an error when the TS signatures are mixed with jsDoc:

```
Error: Command failed with exit code 1: "core/plugins/gutenberg/node_modules/.bin/docgen" packages/core-data/src/selectors.ts --output docs/reference-guides/data/data-core.md --to-token --use-token "Autogenerated selectors|../../../packages/core-data/src/selectors.ts" --ignore "/unstable|experimental/i"

TypeError: Cannot read property '0' of undefined
    at makeError (core/plugins/gutenberg/node_modules/execa/lib/error.js:59:11)
    at handlePromise (core/plugins/gutenberg/node_modules/execa/index.js:114:26)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Transform.<anonymous> (core/plugins/gutenberg/bin/api-docs/update-api-docs.js:219:5) {
  shortMessage: 'Command failed with exit code 1: "core/plugins/gutenberg/node_modules/.bin/docgen" packages/core-data/src/selectors.ts --output docs/reference-guides/data/data-core.md --to-token --use-token "Autogenerated selectors|../../../packages/core-data/src/selectors.ts" --ignore "/unstable|experimental/i"',
  command: '"core/plugins/gutenberg/node_modules/.bin/docgen" packages/core-data/src/selectors.ts --output docs/reference-guides/data/data-core.md --to-token --use-token "Autogenerated selectors|../../../packages/core-data/src/selectors.ts" --ignore "/unstable|experimental/i"',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: "\nTypeError: Cannot read property '0' of undefined",
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
```

I'll just convert all the jsDoc annotations into TypeScript ones, then.

**Comment by adamziel** [comment]
It seems like it's not about mixing the jsDoc annotations with the TypeScript syntax. It's about the TypeScript syntax. I converted the entire selectors.ts file to rely on it, yet the error persists. A quick grep tells me this is the first attempt to use autogenerated annotations with a `.ts` file:

```
$ git grep 'START TOKEN' | grep '\.ts'
docs/reference-guides/data/data-core.md:<!-- START TOKEN(Autogenerated selectors|../../../packages/core-data/src/selectors.ts) -->
packages/core-data/README.md:<!-- START TOKEN(Autogenerated selectors|src/selectors.ts) -->
```

**Comment by dmsnell** [comment]
> A quick grep tells me this is the first attempt to use autogenerated annotations with a .ts file:

That thing keeps giving us more and more trouble; @mirka recently raised an issue for React components wanting to add descriptions to arguments inline.

```ts
/** Does a thing */
function ({
	/** Name of thing */
	name: string,

	/** How many times you want to do the thing. */
	count: number,
}) => { … }
```

cc: @sarayourfriend. I'm not sure what to do because I don't want to divert energy to the docs generation, but I feel like this is going to hit us more and more as we move to TypeScript

**Comment by dmsnell** [comment]
The main issue with `docgen` here is that we've decoupled the descriptions from the types and for wrapped functions that means our parameters are kind of hidden inside the call to the wrapper (`createSelector` and `createRegistrySelector`).

In #40236 I've proposed a hard-coded solution to `docgen` until we can rearchitect it to poll TypeScript for the type information. I think it's worth adding special-cases for selectors in the meantime; that would unblock this PR and other `selectors.ts` changes we'll be making soon.

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
 * @return Record.
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
	): EntityRecord | undefined => {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
export function getUndoEdit( state: State ): Object | undefined {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
export function getRedoEdit( state: State ): any | undefined {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
): EntityRecord | undefined {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
export function getEmbedPreview( state: State, url: string ): any {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
export function getThemeSupports( state: State ): any {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
export function getUndoEdit( state: State ): any | undefined {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
	): EntityRecord | undefined => ( {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
	( state: State, kind: string, name: string, recordId: RecordKey ): any | undefined => {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
): any | undefined {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
): Array< EntityRecord > | undefined {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
	): EntityRecord | undefined => {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
): EntityRecord | null {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
export function getEntity( state: State, kind: string, name: string ): any {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
): Array< any > {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
): Array< any > {
```

File: packages/core-data/src/selectors.ts
Comment (adamziel) [inline packages/core-data/src/selectors.ts]:
```suggestion
): any | null {
```

**Comment by dmsnell** [comment]
I wanted to try and find a way to normalize the types for selectors but I had trouble finding something that was just right, specifically of getting the parameter documentation to work out.

There's a neat trick I stumbled upon which would let us create outwardly-facing parameter names that differ from the internal ones and this has some advantage, but the type signature and syntax is a bit awkward.

```ts
interface State {}

interface Selector<Return, Args extends any[] = []> {
	(state: State, ...args: Args) => Return;
}

// actual types of `state` and `n` are bound by `Selector<…>`
const getBooks: Selector<Book[] | null, [name: string]> = (state, n) => …

// auto-complete shows (state: State, name: string) => …
// but not (state: State, n: string) => …
getBooks(state, 'Roger');
```

The other way is to be more explicit and trap a function type.

```ts
type Selector<F extends ((...args: any[]) => any) | any> =
	F extends (...args: (infer P)) => infer R
		? (state: State, ...args: P) => R
		: (state: State) => F;

const getBooks: Selector<(name: string) => Book[] | null> = (state, n) => …
```

Both of these separate the internal from external names of the function parameters and let us elide the `state` parameter from the outward interface, which given that _all_ selectors _must_ take this as a first parameter seems like a reasonable thing.

Unfortunately I can't seem to tag the parameters with inline documentation and any JSDoc annotations want to only use the internal names from the actual function declaration, not those names from the type declaration.

I didn't realize it before but the way we do this in Simplenote Electron loses the argument information.

Doodling thoughts here. Had hoped to have something a little more exciting to share because I think it would be great if we could say, "If you are writing a selector, use `const mySelector: Selector<ReturnType> = " and then have everything auto-complete.

You are welcome to play with [my exploration of the above](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgEIHt0GtkG8BQyRyYwYANhAFzIDOYUoA5gNyHELrgTg32MhW+AL758oSLEQoACnAZc87IgCNMWWjQzYA2gF02xZCDgBbanQbNDxdADdoAEwCuEVHHJwQSGiGemVaDZRcW4pJGQAZTB5FAIjAAd5KC5NZDkFEH0RMQB6ACp8wnzkACUIMGcoEFpkR3k4ZBgU02QAdzIAC1BkRqZgBxBLdCgIADpi5EmASRhkAE90KroISgQwEeQucnnjCAhHWoADelijkk6eZEo4By3nMF6SeYSUJKgzCugmzaOAQSgTFoRxmczIyFGAEdnMBRrU4I5HGRgFwPL1Af4eGBaotnMhTHBdv07o1yMB6Fs5mBLs9XrV0FTLpM4BjzOB6VBerRaOgEMBYk9HBBaAhGAlSHcTOZ2l1kBBEJ0JsgSpMAALvMxlCpVIYAWi1lWqtJQDJWaw2UCVJXVLM1AKBxH1f0RyNR5HRTEx7JI6DNEHWm0CixAjmQJxikCOABp2t0EJ1kOTkAkUnZgELQ3AHugCaQEB4dlbcvgwC8UJFVv6LQANAA8ADE5QAPSAh2oACnbY27LKBNC8830AEpkABeAB8vRA8xHAB8p-NJ6PlMhGxAWzxDsguz3AWl26AYN8ZEORxPEyAj5zSiujAB+benSA0aKxGPdsa9tInseTm9GIwaHbJ8LFfSAz0nes2HwTgakeECxzwYQuSiCMIGgiRoHgCIK3NEZa3KQ0QEneJiGAtCXzQ99dz7BdhxoQidWCMQYNSR4kkyABhJZeCiSsAygWs-ACaAl0fNCIMsWIxg4lIajGShBGpNhYIpBBENkrhuOccByNiIc2FyXIjAAPTvMRVMech0C8A5tA0F9+ItWt7P0ZB5z8chyDEvTn1Qt9jD4KxBEkkCZOSVIxhgUBHHbBJf2TMYpRQUdUuMIc7zGNRsFqO8H088gVLY64VEcvCoDrdtkqCgQmEk1y9Hc4xnC8nyQJjEBJOs2zHHs2hfIgDqDIs4qVEQ7qQDs9R+va5AAHIABEPAgObhqM0zzNYuDkBUMbRxKgaYzmzjLigVbDOM4gzJG7bopDRCBsogKChKAA5T5KWTCKhjaTp0FoFBso0H0mhisZlWM6rLFqoctHUNyPJa91zwK6C7ti2bFuW1axCAA) if you want, but I think we have to table this for now in favor of the more ad-hoc typings as written in this PR.

**Comment by adamziel** [comment]
```
// actual types of `state` and `n` are bound by `Selector<…>`
const getBooks: Selector<Book[] | null, [name: string]> = (state, n) => …
```

@dmsnell Hm it is quite clever, still I'd rather just call that argument `name`, otherwise I'm not sure what it means as I read the code.

> "If you are writing a selector, use `const mySelector: Selector = " and then have everything auto-complete.

This would be pretty great! And I think we'll get there when integrating the types from `@wordpress/core-data` with the ones from `@wordpress/data`. Perhaps that won't be the first iteration, as you mentioned, but I would really like to make these selectors aware of things like:

* I belong to this store
* Therefore, the state has a certain type
* Also, there are these other selectors I can refer to
* I need to pass the state if I call them directly
* But I don't need to do that when they're thunks

I don't like using these lengthy infer traps, but I believe that's the way here.

File: packages/core-data/README.md
Comment (adamziel) [inline packages/core-data/README.md:291]: I don't like how it says `RecordKey` here, I'd rather have it say `string | number`

File: packages/core-data/README.md
Comment (adamziel) [inline packages/core-data/README.md:490]: It would be helpful to also have these types listed in the documentation.

**Comment by adamziel** [comment]
@dmsnell Yay, all checks are finally green here. Thank you so much for your help with the documentation tools! I'd say this is ready to be reviewed.

File: packages/core-data/README.md
Comment (dmsnell) [reply packages/core-data/README.md:291]: why not? because of the auto-generated documentation?

if that's the case I don't think we can care too much about this problem unless and until we radically change our docs. this is going to be a problem with more and more TypeScript code since this is a win for the types but masks the actual type in the docs. we need hyper-links in our docs to jump to and provide a tooltip description for the types like `RecordKey` here.

unless we get those hyperlinks/tooltips I don't know if there's a good workaround that we should chase.

File: docs/reference-guides/data/data-core.md
Comment (dmsnell) [inline docs/reference-guides/data/data-core.md]: For consistency should we use `RecordKey` here too, and in the other functions expecting a resource `id`?

File: packages/core-data/README.md
Comment (dmsnell) [inline packages/core-data/README.md]: this type doesn't make sense. I know it's supposed to communicate that there might be no return value, but `undefined` is already covered by `any`.

File: packages/core-data/src/selectors.ts
Comment (dmsnell) [inline packages/core-data/src/selectors.ts]: we know this `any` isn't the right type because we later access properties inside of it inside this module, implying it's at least a `Record<string, any>`.

why not create the knowable properties and let it maintain a slightly-higher type value?

This is incomplete but I think suffices for the changes in this patch without introducing knowably-wrong types.

```ts
import type { Context, User, Theme, WpTemplate } from './entity-types';

interface State {
	autosaves: Record< string | number, Array< unknown > >;
	blockPatterns: Array< unknown >;
	blockPatternCategories: Array< unknown >;
	currentGlobalStylesId: string;
	currentTheme: Theme< 'edit' >;
	currentUser: User< 'edit' >;
	embedPreviews: Record< string, { html: string } >;
	entities: EntitiesState;
	themeBaseGlobalStyles: Record< string, Object >;
	themeGlobalStyleVariations: Record< string, string >;
	undo: UndoState;
	users: UserState;
}

interface EntitiesState {
	config: EntityConfig[];
	records: Record< string, unknown >;
}

interface EntityConfig {
	name: string;
	kind: string;
}

interface UndoState extends Array< Object > {
	flattenedUndo: unknown;
	offset: number;
}

interface UserState {
	queries: Record< string, RecordKey[] >;
	byId: Record< RecordKey, User< 'edit' > >;
}
```

File: packages/core-data/src/selectors.ts
Comment (dmsnell) [inline packages/core-data/src/selectors.ts:1085]: is this something we need to update in the docgen tool? is this necessary after the updates in #40236

File: packages/core-data/src/selectors.ts
Comment (dmsnell) [inline packages/core-data/src/selectors.ts]: `null` covered by `any`

File: packages/core-data/src/selectors.ts
Comment (dmsnell) [inline packages/core-data/src/selectors.ts:1148]: ibid.

**Review by dmsnell — COMMENTED** [review]
> it is quite clever, still I'd rather just call that argument name, otherwise I'm not sure what it means as I read the code.

it wasn't the point of the work, just a side-effect. style preferences aside, some languages offer this as a first-class element to the language as a way of giving more appropriate names to parameters from the outside and inside.

for example,

```
function compute( $raw_value: value ) {
    $value = clean( $raw_value );
    …
}
```

Again, not a big deal or the goal, just a side-effect that has its uses.

---

Things mostly look good here. There's a lot of room for improvement which I think is fine to do in follow-up work.

 - Get rid of `Object`
 - Cleanup use of `RecordKey`
 - Expand `State`
 - Audit `any | ?` types.

The `any | ?` types could honestly go in now before merging and it would be good. I get the purpose of `any | null` (or at least I can think of a reasonable use-case) but I don't think it warrants the oddity in the type. A stop-gap could be the use of a nullable type like `Maybe` or `Optional`.

```
type Optional<T> = T | null;
```

And with that we still get `any | null` if we do `Optional<any>` but at least we can see that it was intentional and not an oversight.

The `state` parameter in there for `docgen` is the other thing I'd like to double-check before merge. Is that still necessary? If it is then please go ahead and merge; if it isn't let's get that out before merging.

Either way, thanks for moving on this very large task. Can't wait until we're getting to the even-more exciting bits.

**Review by dmsnell — APPROVED** [review]
Forgot to mark as approved in my submission…

File: packages/core-data/src/selectors.ts
Comment (adamziel) [reply packages/core-data/src/selectors.ts:1148]: What do you mean by ibid here? I know how it's used in bibliographies but I'm having a hard time translating that knowledge to this code review. Are you after wrapping this in an Optional type?

File: packages/core-data/src/selectors.ts
Comment (adamziel) [reply packages/core-data/src/selectors.ts:1085]: Potentially.

Our tooling sees `@param  state Editor state.` in jsdoc but don't see one in the first argument to createSelector because we only use it in the second argument. It would be nice to handle, but that's also kind of a corner case only manifested here. I'm fine with leaving this as is until it comes up someplace else.

File: packages/core-data/src/selectors.ts
Comment (adamziel) [reply packages/core-data/src/selectors.ts]: Good call!

File: packages/core-data/README.md
Comment (adamziel) [reply packages/core-data/README.md:291]: I just don't like how it doesn't tell me what a RecordKey is. How do I construct this type as a user of this code? You nailed it here:

> we need hyper-links in our docs to jump to and provide a tooltip description for the types like RecordKey here.

Once we have that, we're golden. I don't think it needs to be necessarily hard, too. Perhaps there are even existing documentation tools that do that and that wouldn't be too hard to leverage.

**Comment by adamziel** [comment]
> it wasn't the point of the work, just a side-effect. style preferences aside, some languages offer this as a first-class element to the language as a way of giving more appropriate names to parameters from the outside and inside.

Thank you for explaining, I misunderstood your intention. It is a nice feature for sure, and it could come handy for our work here.

Also, I've addressed all your notes, included the additional State types – let's get this one in and address any follow-up work in a follow-up PR.

File: packages/core-data/src/selectors.ts
Comment (dmsnell) [reply packages/core-data/src/selectors.ts:1148]: Didn't mean to complicate things: it means "see the previous comment" or "same comment." Just didn't see the need to write the same comment multiple times.

File: packages/core-data/src/selectors.ts
Comment (dmsnell) [reply packages/core-data/src/selectors.ts:1085]: Ah okay, yeah I see where there's an incompleteness with the `createSelector` exemption. I think we can resolve this and I think the approach in #40865 is reasonable enough.

File: packages/core-data/src/selectors.ts
Comment (dmsnell) [reply packages/core-data/src/selectors.ts:1148]: but I now see that were returning `string | null` so maybe I misread that and thought I had seen `any | null`

File: packages/core-data/src/selectors.ts
Comment (manzoorwanijk) [inline packages/core-data/src/selectors.ts:138]: The context here should be `'view'` instead of `'edit'`, otherwise it will set the wrong types/properties which are not available in current user object, for example, `user.capabilities`.

Comment (manzoorwanijk) [reply packages/core-data/src/selectors.ts:138]: Created #68045 to fix it.

**Pull request merged**
